nav2_assisted_teleop/src/nav2_assisted_teleop.cpp

```diff
@@ -0,0 +1,311 @@
+
+// Copyright (c) 2021 Anushree Sabnis
+// Copyright (c) 2022 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_assisted_teleop/nav2_assisted_teleop.hpp"
+#include "tf2_ros/create_timer_ros.h"
+#include "nav2_util/node_utils.hpp"
+
+#include <thread>
+
+namespace nav2_assisted_teleop
+{
+
+  AssistedTeleop::AssistedTeleop(rclcpp::Logger& logger) : logger_(logger)
+  {}
+
+  void AssistedTeleop::on_configure(std::shared_ptr<nav2_util::LifecycleNode> node,
+                    std::shared_ptr<tf2_ros::Buffer> tf,
+                    std::shared_ptr<nav2_costmap_2d::CostmapTopicCollisionChecker> collision_checker)
+  {
+    // set up parameters
+    node->declare_parameter("transform_tolerance", rclcpp::ParameterValue(0.1));
+
+    nav2_util::declare_parameter_if_not_declared(
+    node, "projection_time", rclcpp::ParameterValue(1.0));
+
+    node->get_parameter("global_frame", global_frame_);
+    node->get_parameter("robot_base_frame", robot_base_frame_);
+
+    // Note: empty frames don't cause robot_utils to error
+    // nav2_util::declare_parameter_if_not_declared(
+    //   node,
+    //   "global_frame", rclcpp::ParameterValue(std::string("global_frame")));
+
+    // nav2_util::declare_parameter_if_not_declared(
+    //   node,
+    //   "robot_base_frame", rclcpp::ParameterValue(std::string("robot_base_frame")));
+
+    tf_ = tf;
+    collision_checker_ = collision_checker;
+  }
+
+  geometry_msgs::msg::Twist AssistedTeleop::computeVelocity(geometry_msgs::msg::Twist& twist)
+  {
+    geometry_msgs::msg::PoseStamped current_pose;
+    if (!nav2_util::getCurrentPose(
+        current_pose, *tf_, global_frame_, robot_base_frame_,
+        transform_tolerance_))
+    {
+      RCLCPP_ERROR(logger_, "Failed to get current pose, setting velocity to zero");
+      return geometry_msgs::msg::Twist();
```

Comment:
I think this should `throw` since that's a critical issue, and in the server, it should catch and 'pass' doing anything in that situation rather than returning an empty message